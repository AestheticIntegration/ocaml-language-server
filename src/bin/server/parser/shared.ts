import * as LSP from "vscode-languageserver-protocol";

type DiagnosticSource = "bucklescript" | "refmt";

export const createDiagnostic = (source: DiagnosticSource) => (
  message: string,
  startCharacter: number,
  startLine: number,
  endCharacter: number,
  endLine: number,
  severity: LSP.DiagnosticSeverity,
) => {
  return {
    code: "",
    message,
    range: {
      end: {
        character: endCharacter,
        line: endLine,
      },
      start: {
        character: startCharacter,
        line: startLine,
      },
    },
    severity,
    source,
  };
};
