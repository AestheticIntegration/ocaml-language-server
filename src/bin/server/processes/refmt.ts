import { ChildProcess } from "child_process";
import * as LSP from "vscode-languageserver-protocol";
import Session from "../session";

export default class ReFMT {
  public readonly process: ChildProcess;
  constructor(session: Session, id?: LSP.TextDocumentIdentifier, argsOpt?: string[]) {
    const uri = id ? id.uri : ".re";
    const command = session.settings.reason.path.refmt;

    const width = session.settings.reason.format.width;
    const widthArg = width === null ? [] : ["--print-width", `${width}`];

    const [cmd, cmdArgs] = session.makeOpamCmd(command);
    const args = argsOpt || ["--parse", "re", "--print", "re", "--interface", `${/\.rei$/.test(uri)}`].concat(widthArg);
    this.process = session.environment.spawn(cmd, cmdArgs.concat(args));
  }
}
