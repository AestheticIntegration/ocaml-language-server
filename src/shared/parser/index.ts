import * as ocamldoc from "./ocamldoc";

export {
  ocamldoc,
};
