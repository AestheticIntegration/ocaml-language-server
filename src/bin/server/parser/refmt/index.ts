import * as LSP from "vscode-languageserver-protocol";
import { createDiagnostic } from "../shared";

const createRefmtDiagnostic = createDiagnostic("refmt");

export function parseErrors(refmtOutput: string): LSP.Diagnostic[] {
  const parsedDiagnostics: LSP.Diagnostic[] = [];

  const reErrors = new RegExp(
    [/File "(.*)", line (\d*), characters (\d*)-(\d*):[\s\S]*?(?:Error|Warning \d+): ([\s\S]*)/]
      .map(r => r.source)
      .join(""),
    "g",
  );

  let errorMatch;
  while ((errorMatch = reErrors.exec(refmtOutput))) {
    // const fileUri = "file://" + errorMatch[1];
    const startLine = Number(errorMatch[2]) - 1;
    const endLine = Number(errorMatch[2]) - 1;
    const startCharacter = Number(errorMatch[3]);
    const endCharacter = Number(errorMatch[4]);
    const message = errorMatch[5].trim();
    const severity = /^Warning number \d+/.exec(errorMatch[0])
      ? LSP.DiagnosticSeverity.Warning
      : LSP.DiagnosticSeverity.Error;

    const diagnostic: LSP.Diagnostic = createRefmtDiagnostic(
      message,
      startCharacter,
      startLine,
      endCharacter,
      endLine,
      severity,
    );
    parsedDiagnostics.push(diagnostic);
  }

  return parsedDiagnostics;
}
