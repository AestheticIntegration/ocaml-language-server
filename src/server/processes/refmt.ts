import { ChildProcess } from "child_process";
import { types } from "../../shared";
import Session from "../session";

export default class ReFMT {
  public readonly process: ChildProcess;
  constructor(session: Session, id?: types.TextDocumentIdentifier, argsOpt?: string[]) {
    const uri = id ? id.uri : ".re";
    const command = session.settings.reason.path.refmt;
    const args = argsOpt || [
      "-use-stdin", "true",
      "-parse", "re",
      "-print", "re",
      "-is-interface-pp", `${/\.rei$/.test(uri)}`,
    ];
    this.process = session.environment.spawn(command, args);
    return this;
  }
}
