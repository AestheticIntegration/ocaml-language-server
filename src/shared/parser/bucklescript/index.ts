import { types } from "../../../shared";

export function parseSyntaxErrors(bsbOutput: string): { [key: string]: types.Diagnostic[] } {
  const parsedDiagnostics = {};

  const reSyntaxErrors = new RegExp ([
    /File "(.*)", line (\d*), characters (\d*)-(\d*):[\s\S]*?/,
    /Error: (.*?)\n[\s\S]*?We've found a bug for you!/,
  ].map((r) => r.source).join(""), "g");

  let errorMatch;
  while (errorMatch = reSyntaxErrors.exec(bsbOutput)) {
    const fileUri = "file://" + errorMatch[1];
    const startLine = Number(errorMatch[2]) - 1;
    const endLine = Number(errorMatch[2]) - 1;
    const startCharacter = Number(errorMatch[3]);
    const endCharacter = Number(errorMatch[4]);
    const message = errorMatch[5].trim();

    const diagnostic: types.Diagnostic = {
      code: "",
      message,
      range: {
        end: {
          character: endCharacter,
          line: endLine,
        },
        start: {
          character: startCharacter,
          line: startLine,
        },
      },
      severity: /^Warning number \d+/.exec(errorMatch[0]) ? types.DiagnosticSeverity.Warning : types.DiagnosticSeverity.Error,
      source: "bucklescript",
    };
    if (!parsedDiagnostics[fileUri]) { parsedDiagnostics[fileUri] = []; }
    parsedDiagnostics[fileUri].push(diagnostic);
  }
  return parsedDiagnostics;
}

export function parseTypeErrors(bsbOutput: string): { [key: string]: types.Diagnostic[] } {
  const parsedDiagnostics = {};
  const reTypeErrors = new RegExp ([
    /(?:We've found a bug for you!|Warning number \d+)\n\s*/, // Heading of the error / warning
    /(.*) (\d+):(\d+)(?:-(\d+)(?::(\d+))?)?\n  \n/, // Capturing file name and lines / indexes
    /(?:.|\n)*?\n  \n/, // Ignoring actual lines content being printed
    /((?:.|\n)*?)/, // Capturing error / warning message
    /(?:\n\S|(?=Warning number \d+)|$)/, // Tail
  ].map((r) => r.source).join(""), "g");

  let errorMatch;
  while (errorMatch = reTypeErrors.exec(bsbOutput)) {
    const fileUri = "file://" + errorMatch[1];
    // Suppose most complex case, path/to/file.re 10:20-15:5 message
    const startLine = Number(errorMatch[2]) - 1;
    const startCharacter = Number(errorMatch[3]) - 1;
    let endLine = Number(errorMatch[4]) - 1;
    let endCharacter = Number(errorMatch[5]); // Non inclusive originally
    const message = errorMatch[6].replace(/\n  /g, "\n");
    if (isNaN(endLine)) {
      // Format path/to/file.re 10:20 message
      endCharacter = startCharacter;
      endLine = startLine;
    } else if (isNaN(endCharacter)) {
      // Format path/to/file.re 10:20-15 message
      endCharacter = endLine + 1; // Format is L:SC-EC
      endLine = startLine;
    }

    const diagnostic: types.Diagnostic = {
      code: "",
      message,
      range: {
        end: {
          character: endCharacter,
          line: endLine,
        },
        start: {
          character: startCharacter,
          line: startLine,
        },
      },
      severity: /^Warning number \d+/.exec(errorMatch[0]) ? types.DiagnosticSeverity.Warning : types.DiagnosticSeverity.Error,
      source: "bucklescript",
    };
    if (!parsedDiagnostics[fileUri]) { parsedDiagnostics[fileUri] = []; }
    parsedDiagnostics[fileUri].push(diagnostic);
  }
  return parsedDiagnostics;
}
