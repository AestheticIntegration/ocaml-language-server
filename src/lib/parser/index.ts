import * as bucklescript from "./bucklescript";
import * as ocamldoc from "./ocamldoc";

export { bucklescript, ocamldoc };
