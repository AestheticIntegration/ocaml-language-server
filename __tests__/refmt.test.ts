import * as fs from "fs";
import * as path from "path";
import * as refmt from "../src/bin/server/parser/refmt";

const mocksFolder = "refmt.mocks";
const getDiagnostics = mockFileName => {
  const bsbOutput = fs.readFileSync(path.join(__dirname, mocksFolder, mockFileName), "utf8");
  return refmt.parseErrors(bsbOutput);
};

it('parses messages including "UNKNOWN SYNTAX" errors', () => {
  expect(getDiagnostics("1.txt")).toMatchSnapshot();
});

it("parses regular syntax error messages", () => {
  expect(getDiagnostics("2.txt")).toMatchSnapshot();
});
