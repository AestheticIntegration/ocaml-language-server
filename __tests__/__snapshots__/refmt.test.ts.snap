// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses messages including "UNKNOWN SYNTAX" errors 1`] = `
Array [
  Object {
    "code": "",
    "message": "2583: <UNKNOWN SYNTAX ERROR>",
    "range": Object {
      "end": Object {
        "character": 74,
        "line": 29,
      },
      "start": Object {
        "character": 73,
        "line": 29,
      },
    },
    "severity": 1,
    "source": "refmt",
  },
]
`;

exports[`parses regular syntax error messages 1`] = `
Array [
  Object {
    "code": "",
    "message": "2146: syntax error, consider adding a \`;' before",
    "range": Object {
      "end": Object {
        "character": 22,
        "line": 41,
      },
      "start": Object {
        "character": 7,
        "line": 41,
      },
    },
    "severity": 1,
    "source": "refmt",
  },
]
`;
